added error when executable crashes

lynnmunday opened the pull request:
closes #30340

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2822258/), step Docs: sync website on 7ad2b12 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30341/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2822273/), step Generate coverage on 7ad2b12 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30341/coverage/xfem)

</details>



This comment will be updated on new commits.

dschwen commented on python/TestHarness/util.py:
```diff
@@ -528,7 +528,12 @@ def getCapabilities(exe):
     """
     assert exe
-    output = runCommand("%s --show-capabilities" % exe)
+    try:
+        output = runCommand("%s --show-capabilities" % exe)
+    except:
+        print("Error! Executable crashed.")
+        exit(1)
+
```
You could put this `try`/`catch` into `runCommand`...

lynnmunday replied:
It will change what people see, instead of seeing a bunch of test failed because it crashed, they will only see:

Error! Executable crashed.

lynnmunday replied:
I was wrong, I wasn't catching the exception.  I pasted in the error messages below.  I added some extra output to the exception that has always been raised.

lynnmunday commented:
This is was the old error message I got when the executable segfaults with --show-capabilities
```
(moose) INL621590:porous_flow mundlb$ rt
Traceback (most recent call last):
  File "/Users/mundlb/projects/moose/python/TestHarness/util.py", line 716, in parseMOOSEJSON
    output = output.split('**START JSON DATA**\n')[1]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mundlb/projects/moose/modules/porous_flow/./run_tests", line 8, in <module>
    TestHarness.buildAndRun(sys.argv, None, MOOSE_DIR)
  File "/Users/mundlb/projects/moose/python/TestHarness/TestHarness.py", line 231, in buildAndRun
    harness = TestHarness(argv, moose_dir, moose_python_dir, test_root)
  File "/Users/mundlb/projects/moose/python/TestHarness/TestHarness.py", line 311, in __init__
    self.options._capabilities = util.getCapabilities(self.executable)
  File "/Users/mundlb/projects/moose/python/TestHarness/util.py", line 532, in getCapabilities
    return parseMOOSEJSON(output, '--show-capabilities')
  File "/Users/mundlb/projects/moose/python/TestHarness/util.py", line 724, in parseMOOSEJSON
    raise Exception(output_string)
Exception:  Failed to find JSON header and footer from --show-capabilities
```

With this change it will show any errors the executable might have produced and the exception:
```
(moose) INL621590:porous_flow mundlb$ rt
Traceback (most recent call last):
  File "/Users/mundlb/projects/moose/python/TestHarness/util.py", line 716, in parseMOOSEJSON
    output = output.split('**START JSON DATA**\n')[1]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mundlb/projects/moose/modules/porous_flow/./run_tests", line 8, in <module>
    TestHarness.buildAndRun(sys.argv, None, MOOSE_DIR)
  File "/Users/mundlb/projects/moose/python/TestHarness/TestHarness.py", line 231, in buildAndRun
    harness = TestHarness(argv, moose_dir, moose_python_dir, test_root)
  File "/Users/mundlb/projects/moose/python/TestHarness/TestHarness.py", line 311, in __init__
    self.options._capabilities = util.getCapabilities(self.executable)
  File "/Users/mundlb/projects/moose/python/TestHarness/util.py", line 532, in getCapabilities
    return parseMOOSEJSON(output, '--show-capabilities')
  File "/Users/mundlb/projects/moose/python/TestHarness/util.py", line 724, in parseMOOSEJSON
    raise Exception(output_string)
Exception:
ERROR: dyld[67113]: symbol not found in flat namespace '__ZN39PorousFlowPermeabilityKozenyCarmanTemplILb0EE19computeQpPropertiesEv'
Failed to find JSON header and footer from --show-capabilities
```

I think this would have helped me find my error.

moosebuild commented:
Job [Python 3.11](https://civet.inl.gov/job/2826922/) on 61de797 : invalidated by @lynnmunday

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2826933/) on 61de797 : invalidated by @lynnmunday

navierstokes distributed failed

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2826951/), step Verify coverage on 61de797 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `subchannel` coverage rate (required: 87.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2826951/), step Verify coverage on 61de797 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `subchannel` coverage rate (required: 87.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

lynnmunday commented:
@dschwen Will you review and merge?

lynnmunday commented:
The latest commit will only show:
```
(moose) INL621590:poroperm mundlb$ rt
Error! Executable crashed running --show-capbilities.
```

lynnmunday committed:
added error when executable crashes with the --show-capabilities option

closes #30340

dschwen commented on python/TestHarness/util.py:
```diff
@@ -529,6 +529,9 @@ def getCapabilities(exe):
     """
     assert exe
     output = runCommand("%s --show-capabilities" % exe)
+    if output.startswith("ERROR"):
+        print("Error! Executable crashed running --show-capbilities.")
+        exit(1)
```
No, that's not right. You might get a crash without the `ERROR` string.

lynnmunday replied:
This is right, you get an ERROR whenever the executable crashes.

line 226 appends ERROR when the executable crashes.
```
    p = subprocess.Popen(cmd, cwd=cwd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, close_fds=should_close, shell=True)
    output = p.communicate()[0].decode('utf-8')
    if (p.returncode != 0):
        output = 'ERROR: ' + output
```
I can't change the code in the runCommand function because its called by the testHarness and shouldn't exit if it segfaults, it should return an ERROR.
